src/coreclr/src/jit/liveness.cpp

```diff
@@ -2109,40 +2109,80 @@ void Compiler::fgComputeLifeLIR(VARSET_TP& life, BasicBlock* block, VARSET_VALAR
                 break;
 
             case GT_NOP:
+            {
                 // NOTE: we need to keep some NOPs around because they are referenced by calls. See the dead store
                 // removal code above (case GT_STORE_LCL_VAR) for more explanation.
                 if ((node->gtFlags & GTF_ORDER_SIDEEFF) != 0)
                 {
                     break;
                 }
-                __fallthrough;
+                fgTryRemoveNonLocal(node, &blockRange);
+            }
+            break;
 
-            default:
-                assert(!node->OperIsLocal());
-                if (!node->IsValue() || node->IsUnusedValue())
+            case GT_BLK:
+            case GT_OBJ:
+            case GT_DYN_BLK:
+            {
+                bool removed = fgTryRemoveNonLocal(node, &blockRange);
+                if (!removed && node->IsUnusedValue())
                 {
-                    // We are only interested in avoiding the removal of nodes with direct side-effects
-                    // (as opposed to side effects of their children).
-                    // This default case should never include calls or assignments.
-                    assert(!node->OperRequiresAsgFlag() && !node->OperIs(GT_CALL));
-                    if (!node->gtSetFlags() && !node->OperMayThrow(this))
-                    {
-                        JITDUMP("Removing dead node:\n");
-                        DISPNODE(node);
-
-                        node->VisitOperands([](GenTree* operand) -> GenTree::VisitResult {
-                            operand->SetUnusedValue();
-                            return GenTree::VisitResult::Continue;
-                        });
-
-                        blockRange.Remove(node);
-                    }
+                    // IR doesn't expect dummy uses of `GT_OBJ/BLK/DYN_BLK`.
+                    node->ChangeType(TYP_BYTE);
+                    node->ChangeOper(GT_NULLCHECK);
+                    block->bbFlags |= BBF_HAS_NULLCHECK;
+                    optMethodFlags |= OMF_HAS_NULLCHECK;
```

Comment:
We should probably factor this to a helper, e.g., `ChangeOperToNullCheck` but it can be done in a separate PR.

Reply:
I don't think this matters, but `Lowering` may change these to `GT_IND` because that produces better code for some cases.

Reply:
Yes, I saw that code, unconsumed `IND` are fine, we had asserts only about `OBJ/BKJ/DYN_BLK`.

Reply:
Right - my point was that it might be nice to avoid having to transform multiple times, but neither do we want to duplicate the determination of which form to use.